working/0546-patterns/patterns-feature-specification.md

```diff
@@ -541,24 +540,40 @@ is expected.
 ### List pattern
 
 ```
-listPattern ::= typeArguments? '[' patterns? ']'
+listPattern         ::= typeArguments? '[' listPatternElements? ']'
+listPatternElements ::= listPatternElement ( ',' listPatternElement )* ','?
+listPatternElement  ::= pattern | restPattern
+restPattern         ::= '...' pattern?
 ```
 
 A list pattern matches an object that implements `List` and extracts elements by
 position from it.
 
-It is a compile-time error if `typeArguments` is present and has more than one
-type argument.
+It is a compile-time error if:
+
+*   `typeArguments` is present and has more than one type argument.
+
+*   There is more than one `restPattern` element in the list pattern. *It can
+    appear anywhere in the list, but there can only be zero or one.*
 
-**TODO: Allow a `...` element in order to match suffixes or ignore extra
-elements. Allow capturing the rest in a variable.**
+#### Rest patterns
+
+A list pattern may contain a *rest element* which allows matching lists of
+arbitrary lengths. If a rest element is present and has a subpattern, all of the
+elements not matched by other subpatterns are collected into a new list and that
+list is matched against the rest subpattern.
+
+```dart
+var [a, b, ...rest, c, d] = [1, 2, 3, 4, 5, 6, 7];
```

Comment:
Should the pattern have a required type of `List<S>` for some kind of `S`?  That is, I think in a refutable context we would otherwise allow `var [... String rest]` which can't ever match, right?  Or is that an error for other reasons?

Reply:
> I think in a refutable context we would otherwise allow `var [... String rest]` which can't ever match, right?

That's true, but only reliably true because we happen to know that String is sealed. I think this is similar to the general property where we don't report an error when a subpattern requires a type unrelated to the matched value type because there could be some unknown third type that is a subtype of both. It might be weird to have:

```dart
var [...Future foo] = ...
```

But some particularly irascible Dart programmer *could* define:

```dart
class Wat<T> implements List<T>, Future<T> { ... }
```

So even though `[...String rest]` is weird and useless, I don't know if it needs to be an error.

Reply:
I'm pretty confused now.  What is your intended semantics of this?

```
var [...String x] = <int>[3];
```

I had expected that you would say that the required type for `String x` is `List<int>`, and that this would be an error.  But you seem to be saying that no, this is fine, there's no required type for the pattern.  In which case.... this pattern is refutable?  Or is there some other static error that applies here that I'm missing?

Reply:
The required type of `String x` is `String`. The matched value type is `List<int>`. Since `List<int>` is not assignable to `String`, that should be a compile-time error.
The required type of a pattern is the type that it can actually match. It only depends on the matched value type when the pattern is underspecified, and we infer its type through type inference from the matched value type.

In a refutable context, the `[...String x]` will just never match, but neither will `[Never x]`, and `someList is String` will always be false. We don't generally prevent you from doing impossible checks (but the analyzer is free to warn you if you do).

Reply:
Lasse is correct. Sorry, I got my example wrong. In your previous comment, you mentioned "in a refutable context" (which is correct), but then my example was an irrefutable use of a pattern. I should have written:

```dart
case [...Future foo]: ...
```

This doesn't have a compile-time error. It's theoretically possible that the `List` returned by the `sublist()` call on the destructured list happens to return some class that both implements `List` and `Future`, so we don't prohibit it outright. But this *is* a compile-time error:

```dart
var [...Future foo] = ...
```

Because we don't know that it *will* match, in an irrefutable context, we prohibit this at compile-time since pattern variable declarations should never fail to match.

There is an interesting wrinkle here. Someone could in theory define:

```dart
class FunnyList implements List<int> {
  ListAndFuture sublist() { ... }
  ...
}

class ListAndFuture implements List<int>, Future { ... }
```

Then if you were to do:

```dart
var [...Future f] = FunnyList();
```

We *could* allow this. The question is whether the matched value type of a rest subpattern is just `List<E>` where `E` is the surrounding list pattern's required type's element type, or we could say that the matched value type is the return type of the `sublist()` method on the matched value of the surrounding list. The proposal says the former, but it would potentially be more precise to pick the latter because then `sublist()` methods with covariant returns could return a more precise type.

In practice... I don't think anyone will do that, so I'm OK with the proposal where it is.

Reply:
If we define the matched value type of a list spread pattern to be the same list type as the outer list pattern, and don't try to be clever and take the return type of `sublist`, then the irrefutable example will be invalid.

It won't stop the same example in a refutable context from matching, though. I guess that's OK. 
I expect a lint warning about pattern match between unrelated types, and the analyzer warnings about non-list types in spread patterns.